net: Fix out-of-bounds access when registering connections

lpereira opened the pull request:
If either a remote or a local address were supplied to the
net_conn_register() function, the IP stack would proceed to copy
sizeof(struct sockaddr) bytes from the respective remote_addr
or local_addr pointers, regardless of the actual size of the storage
these pointers point to.

Use the proper size depending on the socket address family.

Coverity-ID: 173630
Signed-off-by: Leandro Pereira <leandro.pereira@intel.com>

lpereira committed:
net: Fix out-of-bounds access when registering connections

If either a remote or a local address were supplied to the
net_conn_register() function, the IP stack would proceed to copy
sizeof(struct sockaddr) bytes from the respective remote_addr
or local_addr pointers, regardless of the actual size of the storage
these pointers point to.

Use the proper size depending on the socket address family.

Coverity-ID: 173630
Signed-off-by: Leandro Pereira <leandro.pereira@intel.com>

mike-scott commented:
If we're not clearing out the rest of the sockaddr structure, will that cause issues somewhere else in the code?  I don't think it is since most check the sa_family and use the data appropriately.. but I wanted to bring it up as a discussion point.  (Not that we're going to flip/flop remote or local address a bunch).

lpereira commented:
It's been a while since I've touched the networking layer, but it should check the `sa_family_t` member every time it's going to cast to the proper address structure.

I suppose `conn[i].{remote,local}_addr` could be zeroed out before copying the address passed from the user, to avoid surprises?

mike-scott commented:
Yep, still not sure if it's necessary.  But might be prudent.

lpereira committed:
net: connection: Zero out connection struct when unregistering

Zeroing-out all bytes from a connection structure when it's
unregistered ensures all state from a previous connection is gone and
can't be mistakenly reused.

Signed-off-by: Leandro Pereira <leandro.pereira@intel.com>

lpereira commented:
@mike-scott Instead of zeroing out only the address structs, the whole connection struct is zeroed out when the connection is unregistered.  It should go back to the state it was since boot, since the conns array is a static global.

mike-scott reviewed (approved):
LGTM

jukkar commented:
Currently both the "struct sockaddr" and "struct sockaddr_in6" are the same size if IPv6 is enabled. So atm there should be no issue with size. This is different than what you see when working in linux where the size differs. Anyway it is good that this is made more strict so we do not get any surprises if we decide to change the size of the "struct sockaddr".
